// @symblai/symbl-js/dist/event/StopProcessingEvent.js
"use strict";
exports.__esModule = true;
var StopProcessingEvent = /** @class */ (function () {
    function StopProcessingEvent(options) {
        if (options === void 0) { options = {}; }
        if (!options) {
            throw new Error('StopProcessing configuration required');
        }
        this._timestamp = options.timestamp;
        this._topic = options.topic;
        this._type = 'stopped';
    }
    StopProcessingEvent.topics = function () {
        return {
            speaker: 'speaker'
        };
    };
    StopProcessingEvent.types = function () {
        return {
            stopped: 'stopped'
        };
    };
    Object.defineProperty(StopProcessingEvent.prototype, "timestamp", {
        get: function () {
            return this._timestamp;
        },
        set: function (val) {
            this._timestamp = val;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StopProcessingEvent.prototype, "type", {
        get: function () {
            return this._type;
        },
        set: function (val) {
            this._type = val;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StopProcessingEvent.prototype, "topic", {
        get: function () {
            return this._topic;
        },
        set: function (val) {
            this._topic = val;
        },
        enumerable: false,
        configurable: true
    });
    StopProcessingEvent.prototype.toJSON = function () {
        return {
            topic: this._topic,
            type: this._type,
            timestamp: this._timestamp
        };
    };
    return StopProcessingEvent;
}());
exports["default"] = StopProcessingEvent;
;
