// @symblai/api-client/src/model/Endpoint.js
/**
 * symbl.ai APIs
 * <h1>Symbl's APIs for external consumers.</h1> <h2>Language Insights API</h2> Symbl's Language Insights API provides an interface for applications to perform the analysis on the raw text and get insights from it. The API automatically detects sentence boundaries and punctuates the sentences, and also returns the updated messages in the response. Conversations are the most unstructured piece of information that we represent information in, and which most of the leads to lot of loss of information by not being able to capture them correctly.<br/> Language Insights API focuses on understanding such texts and generate the useful and important information from them. <br/> Currently the API supports detection of the Action Items in any type of unstructured text. In future the same API will also have support to detect \"Information\" and \"Event\", where Information is any informational piece and Event is a reference to something that has happened in the past.<br/> <h2>Telephony Integration</h2> Symbl can currently integrate with two types of telephony endpoints: 1. SIP trunks<br/> 2. PSTN endpoints<br/> Results are sent via HTTP WebHooks as and when they are available.<br/> <h2>Flow</h2> 1. External Application invokes REST API to join a meeting/session, with the mode (SIP/PSTN) and joining details<br/> 2. Symbl joins the meeting via SIP or PSTN integration<br/> 3. Symbl continuously processes the audio stream received<br/> 4. Symbl calls WebHook whenever transcription results are available<br/>
 *
 * OpenAPI spec version: 1.0
 * Contact: info@symbl.ai
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 2.3.1
 *
 * Do not edit the class manually.
 *
 */

(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient', 'model/AudioConfig'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('./AudioConfig'));
  } else {
    // Browser globals (root is window)
    if (!root.SymblaiApIs) {
      root.SymblaiApIs = {};
    }
    root.SymblaiApIs.Endpoint = factory(root.SymblaiApIs.ApiClient, root.SymblaiApIs.AudioConfig);
  }
}(this, function(ApiClient, AudioConfig) {
  'use strict';




  /**
   * The Endpoint model module.
   * @module model/Endpoint
   * @version 1.0
   */

  /**
   * Constructs a new <code>Endpoint</code>.
   * @alias module:model/Endpoint
   * @class
   */
  var exports = function() {
    var _this = this;











  };

  /**
   * Constructs a <code>Endpoint</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/Endpoint} obj Optional instance to populate.
   * @return {module:model/Endpoint} The populated <code>Endpoint</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      if (data.hasOwnProperty('type')) {
        obj['type'] = ApiClient.convertToType(data['type'], 'String');
      }
      if (data.hasOwnProperty('uri')) {
        obj['uri'] = ApiClient.convertToType(data['uri'], 'String');
      }
      if (data.hasOwnProperty('providerName')) {
        obj['providerName'] = ApiClient.convertToType(data['providerName'], 'String');
      }
      if (data.hasOwnProperty('phoneNumber')) {
        obj['phoneNumber'] = ApiClient.convertToType(data['phoneNumber'], 'String');
      }
      if (data.hasOwnProperty('dtmf')) {
        obj['dtmf'] = ApiClient.convertToType(data['dtmf'], 'String');
      }
      if (data.hasOwnProperty('language')) {
        obj['language'] = ApiClient.convertToType(data['language'], 'String');
      }
      if (data.hasOwnProperty('audioConfig')) {
        obj['audioConfig'] = AudioConfig.constructFromObject(data['audioConfig']);
      }
      if (data.hasOwnProperty('timezoneOffset')) {
        obj['timezoneOffset'] = ApiClient.convertToType(data['timezoneOffset'], 'Number');
      }
      if (data.hasOwnProperty('idleTimeout')) {
        obj['idleTimeout'] = ApiClient.convertToType(data['idleTimeout'], 'Number');
      }
      if (data.hasOwnProperty('transportConfig')) {
        obj['transportConfig'] = ApiClient.convertToType(data['transportConfig'], 'String');
      }
    }
    return obj;
  }

  /**
   * @member {module:model/Endpoint.TypeEnum} type
   */
  exports.prototype['type'] = undefined;
  /**
   * SIP URI to connect to
   * @member {String} uri
   */
  exports.prototype['uri'] = undefined;
  /**
   * The name of the meeting platform to connect to
   * @member {String} providerName
   */
  exports.prototype['providerName'] = undefined;
  /**
   * Phone number to be used to dial in to in E.164 format i.e. special characters like () or - and leading + or international access codes like 001 or 00 must be omitted. For e.g. - US number should look like 14082924837, whereas UK number should look like 447082924837
   * @member {String} phoneNumber
   */
  exports.prototype['phoneNumber'] = undefined;
  /**
   * DTMF sequence to be sent after call is received
   * @member {String} dtmf
   */
  exports.prototype['dtmf'] = undefined;
  /**
   * Language code as per the BCP 47 specification
   * @member {String} language
   */
  exports.prototype['language'] = undefined;
  /**
   * The audio config specifing audio stream specific configurations for the call.
   * @member {module:model/AudioConfig} audioConfig
   */
  exports.prototype['audioConfig'] = undefined;
  /**
   * Number of minutes that needs to be added or removed from the UTC time. Positive or negative numbers are accepted. For example, for PST timezone, the value should be -480 and for IST timezone, the value should be 330
   * @member {Number} timezoneOffset
   */
  exports.prototype['timezoneOffset'] = undefined;
  /**
   * The timeout value in seconds for Symbl to leave the meeting if no one is speaking. -1 for no timeout which is the default value if not specified.
   * @member {Number} idleTimeout
   */
  exports.prototype['idleTimeout'] = undefined;
  /**
   * The transport-layer protocols name and value pairs separated by semi-colons to use for making the SIP requests
   * @member {String} transportConfig
   */
  exports.prototype['transportConfig'] = undefined;


  /**
   * Allowed values for the <code>type</code> property.
   * @enum {String}
   * @readonly
   */
  exports.TypeEnum = {
    /**
     * value: "sip"
     * @const
     */
    "sip": "sip",
    /**
     * value: "pstn"
     * @const
     */
    "pstn": "pstn"  };


  return exports;
}));
