// @symblai/api-client/src/model/UserTrunc.js
/**
 * symbl.ai APIs
 * <h1>Symbl's APIs for external consumers.</h1> <h2>Language Insights API</h2> Symbl's Language Insights API provides an interface for applications to perform the analysis on the raw text and get insights from it. The API automatically detects sentence boundaries and punctuates the sentences, and also returns the updated messages in the response. Conversations are the most unstructured piece of information that we represent information in, and which most of the leads to lot of loss of information by not being able to capture them correctly.<br/> Language Insights API focuses on understanding such texts and generate the useful and important information from them. <br/> Currently the API supports detection of the Action Items in any type of unstructured text. In future the same API will also have support to detect \"Information\" and \"Event\", where Information is any informational piece and Event is a reference to something that has happened in the past.<br/> <h2>Telephony Integration</h2> Symbl can currently integrate with two types of telephony endpoints: 1. SIP trunks<br/> 2. PSTN endpoints<br/> Results are sent via HTTP WebHooks as and when they are available.<br/> <h2>Flow</h2> 1. External Application invokes REST API to join a meeting/session, with the mode (SIP/PSTN) and joining details<br/> 2. Symbl joins the meeting via SIP or PSTN integration<br/> 3. Symbl continuously processes the audio stream received<br/> 4. Symbl calls WebHook whenever transcription results are available<br/>
 *
 * OpenAPI spec version: 1.0
 * Contact: info@symbl.ai
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 2.3.1
 *
 * Do not edit the class manually.
 *
 */

(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient', 'model/Avatar', 'model/UserTrunc'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('./Avatar'), require('./UserTrunc'));
  } else {
    // Browser globals (root is window)
    if (!root.SymblaiApIs) {
      root.SymblaiApIs = {};
    }
    root.SymblaiApIs.UserTrunc = factory(root.SymblaiApIs.ApiClient, root.SymblaiApIs.Avatar, root.SymblaiApIs.UserTrunc);
  }
}(this, function(ApiClient, Avatar, UserTrunc) {
  'use strict';




  /**
   * The UserTrunc model module.
   * @module model/UserTrunc
   * @version 1.0
   */

  /**
   * Constructs a new <code>UserTrunc</code>.
   * @alias module:model/UserTrunc
   * @class
   * @param name {String} Name of the user
   * @param userId {String} Unique Identifier for user, recommended to be human readable. Usually this should be username or email Id etc. that user is also aware of.
   */
  var exports = function(name, userId) {
    var _this = this;


    _this['name'] = name;
    _this['userId'] = userId;




  };

  /**
   * Constructs a <code>UserTrunc</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/UserTrunc} obj Optional instance to populate.
   * @return {module:model/UserTrunc} The populated <code>UserTrunc</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      if (data.hasOwnProperty('id')) {
        obj['id'] = ApiClient.convertToType(data['id'], 'String');
      }
      if (data.hasOwnProperty('name')) {
        obj['name'] = ApiClient.convertToType(data['name'], 'String');
      }
      if (data.hasOwnProperty('userId')) {
        obj['userId'] = ApiClient.convertToType(data['userId'], 'String');
      }
      if (data.hasOwnProperty('preferredGender')) {
        obj['preferredGender'] = ApiClient.convertToType(data['preferredGender'], 'String');
      }
      if (data.hasOwnProperty('role')) {
        obj['role'] = ApiClient.convertToType(data['role'], 'String');
      }
      if (data.hasOwnProperty('avatar')) {
        obj['avatar'] = Avatar.constructFromObject(data['avatar']);
      }
      if (data.hasOwnProperty('additionalUsers')) {
        obj['additionalUsers'] = ApiClient.convertToType(data['additionalUsers'], [UserTrunc]);
      }
    }
    return obj;
  }

  /**
   * Optional Identifier other than userId to identify the user
   * @member {String} id
   */
  exports.prototype['id'] = undefined;
  /**
   * Name of the user
   * @member {String} name
   */
  exports.prototype['name'] = undefined;
  /**
   * Unique Identifier for user, recommended to be human readable. Usually this should be username or email Id etc. that user is also aware of.
   * @member {String} userId
   */
  exports.prototype['userId'] = undefined;
  /**
   * Optional. Preferred gender of the user.
   * @member {module:model/UserTrunc.PreferredGenderEnum} preferredGender
   */
  exports.prototype['preferredGender'] = undefined;
  /**
   * String representing any role that this user may have
   * @member {String} role
   */
  exports.prototype['role'] = undefined;
  /**
   * @member {module:model/Avatar} avatar
   */
  exports.prototype['avatar'] = undefined;
  /**
   * Users in addition to the primary user. This is useful in the message authors to capture overlapping speakers.
   * @member {Array.<module:model/UserTrunc>} additionalUsers
   */
  exports.prototype['additionalUsers'] = undefined;


  /**
   * Allowed values for the <code>preferredGender</code> property.
   * @enum {String}
   * @readonly
   */
  exports.PreferredGenderEnum = {
    /**
     * value: "male"
     * @const
     */
    "male": "male",
    /**
     * value: "female"
     * @const
     */
    "female": "female",
    /**
     * value: "unknown"
     * @const
     */
    "unknown": "unknown"  };


  return exports;
}));
